common/solver/batch_gmres_kernels.hpp.inc

```diff
@@ -0,0 +1,985 @@
+        const uint32 conv = converged & (1 << c);
+
+        if (conv) {
+            continue;
+        }
+
+        if (y[c] == zero<ValueType>()) {
+            cs[c] = one<ValueType>();
+            sn[c] = zero<ValueType>();
+        } else if (abs(y[c]) > abs(x[c])) {
+            ValueType temp = x[c] / y[c];
+            sn[c] = one<ValueType>() / sqrt(one<ValueType>() + temp * temp);
+            cs[c] = temp * sn[c];
+        } else {
+            ValueType temp = y[c] / x[c];
+            cs[c] = one<ValueType>() / sqrt(one<ValueType>() + temp * temp);
+            sn[c] = temp * cs[c];
+        }
+    }
+}
+
+template <typename ValueType>
+__device__ __forceinline__ void ApplyPlaneRotation(
+    ValueType *const x, ValueType *const y, const ValueType *const cs,
+    const ValueType *const sn, const int nrhs, const uint32 &converged)
+{
+    for (int c = threadIdx.x; c < nrhs; c += blockDim.x) {
+        const uint32 conv = converged & (1 << c);
+
+        if (conv) {
+            continue;
+        }
+
+        ValueType temp = cs[c] * x[c] + sn[c] * y[c];
+        y[c] = -one<ValueType>() * sn[c] * x[c] + cs[c] * y[c];
+        x[c] = temp;
+    }
+}
+
+
+template <typename ValueType>
+__device__ __forceinline__ void update_x(
+    const gko::batch_dense::BatchEntry<ValueType> &x_shared_entry, const int m,
+    const gko::batch_dense::BatchEntry<const ValueType> &H_shared_entry,
+    const gko::batch_dense::BatchEntry<const ValueType> &s_shared_entry,
+    const gko::batch_dense::BatchEntry<const ValueType> &V_shared_entry,
+    const uint32 &converged)
+{
+    __shared__
+        UninitializedArray<ValueType, gko::kernels::batch_gmres::max_restart *
+                                          batch_config<ValueType>::max_num_rhs>
+            y_shared_sh;
+
+    const gko::batch_dense::BatchEntry<ValueType> y_shared_entry{
+        y_shared_sh, static_cast<size_type>(x_shared_entry.num_rhs), m + 1,
+        x_shared_entry.num_rhs};
+
+
+    // triangular solve
+
+
+    for (int rhs_index = threadIdx.x; rhs_index < y_shared_entry.num_rhs;
+         rhs_index += blockDim.x) {
+        for (int r = m; r >= 0; r--) {
+            ValueType temp_sum = zero<ValueType>();
+
+            for (int t = m; t > r; t--) {
+                temp_sum +=
+                    H_shared_entry
+                        .values[r * H_shared_entry.stride +
+                                t * y_shared_entry.num_rhs + rhs_index] *
+                    y_shared_entry
+                        .values[t * y_shared_entry.stride + rhs_index];
+            }
+
+            y_shared_entry.values[r * y_shared_entry.stride + rhs_index] =
+                (s_shared_entry.values[r * s_shared_entry.stride + rhs_index] -
+                 temp_sum) /
+                H_shared_entry.values[r * H_shared_entry.stride +
+                                      r * y_shared_entry.num_rhs + rhs_index];
+        }
+    }
+
+    __syncthreads();
+
```

Comment:
Like my comment for the reference code, I think it would be good to split this function into the triangular solve separate from the update of `x`. That way, we would not be repeating the triangular solve at the end of each restart cycle. That may be somewhat more important here because we expect short restart cycles.